File: android-stery-print-bridge/app/src/main/java/com/stery/bluetoothprintbridge/ReceiptFormatter.kt
package com.stery.bluetoothprintbridge

import android.content.Context
import android.graphics.BitmapFactory
import android.util.Log
import java.io.ByteArrayOutputStream
import java.util.Locale

class ReceiptFormatter(
    private val context: Context? = null,
    private val width: Int = 32,
) {

    fun format(payload: ReceiptPayload): ByteArray {
        val paymentStatus = payload.paymentStatus?.trim().orEmpty()
        val balanceDue = payload.balance ?: maxOf((payload.total ?: 0.0) - (payload.paid ?: 0.0), 0.0)
        val isCreditReceipt =
            payload.isCredit ||
                payload.saleType.equals("credit", ignoreCase = true) ||
                paymentStatus.contains("credit", ignoreCase = true) ||
                balanceDue > 0.0 ||
                (payload.total ?: 0.0) > (payload.paid ?: 0.0)
        Log.d(
            "PRINT",
            "ANDROID FORMAT CREDIT isCreditReceipt=$isCreditReceipt isCredit=${payload.isCredit} saleType=${payload.saleType} paymentStatus=${payload.paymentStatus} balance=${payload.balance} paid=${payload.paid}"
        )

        val out = ByteArrayOutputStream()
        emit(out, EscPos.init())
        emit(out, EscPos.alignCenter())
        emit(out, EscPos.bold(true))
        payload.duplicateLabel?.takeIf { it.isNotBlank() }?.let { emitLineCentered(out, it) }
        emit(out, EscPos.doubleSize(true))
        if (payload.printLogo) {
            renderLogo(out)
        }
        emitLineCentered(out, payload.storeName)
        emit(out, EscPos.doubleSize(false))
        emit(out, EscPos.bold(false))
        payload.branch?.takeIf { it.isNotBlank() }?.let { emitLineCentered(out, it) }
        emitLineCentered(out, "SALE RECEIPT")
        emit(out, EscPos.alignLeft())
        if (!payload.receiptNo.isNullOrBlank()) emitKeyValue(out, "Receipt", payload.receiptNo)
        if (!payload.date.isNullOrBlank()) emitKeyValue(out, "Date", payload.date)
        val cashierName = payload.cashierName ?: payload.cashier
        val customerName = payload.customerName ?: payload.customer
        val deliveryPerson = payload.deliveryPersonName ?: payload.deliveryPerson
        val effectiveConditions = if (isCreditReceipt) {
            payload.conditions + "Accounts are due on demand and overdue accounts attract interest at 3% per week."
        } else {
            payload.conditions
        }
        if (!cashierName.isNullOrBlank()) emitKeyValue(out, "Cashier", cashierName)
        if (!customerName.isNullOrBlank()) emitKeyValue(out, "Customer", customerName)
        if (!payload.saleType.isNullOrBlank()) emitKeyValue(out, "Sale Type", payload.saleType)
        if (isCreditReceipt && !deliveryPerson.isNullOrBlank()) emitKeyValue(out, "Delivery Person", deliveryPerson)
        if (isCreditReceipt) {
            emitCenteredBoldLine(out, "CREDIT SALE")
            if (!paymentStatus.isNullOrBlank() && !paymentStatus.equals("paid", ignoreCase = true)) emitKeyValue(out, "Status", paymentStatus)
            if (balanceDue > 0.0) emitBoldKeyValue(out, "BALANCE DUE", money(balanceDue))
        } else if (!paymentStatus.isNullOrBlank() && !paymentStatus.equals("paid", ignoreCase = true)) {
            emitKeyValue(out, "Status", paymentStatus)
        }
        payload.note?.takeIf { it.isNotBlank() }?.let {
            emitDivider(out)
            emitWrappedBlock(out, it)
        }

        emitDivider(out)
        emit(out, EscPos.bold(true))
        emitLine(out, leftRight("Item / Price x Qty", "Amt"))
        emit(out, EscPos.bold(false))
        emitDivider(out)

        payload.items.forEach { item ->
            formatItem(item).forEach { emitLine(out, it) }
        }

        emitDivider(out)
        payload.subtotal?.let { emitKeyValue(out, "Subtotal", money(it)) }
        payload.discount?.let { emitKeyValue(out, "Discount", money(it)) }
        (payload.vat ?: payload.tax)?.let { emitKeyValue(out, "VAT", money(it)) }
        payload.netAmount?.let { emitKeyValue(out, "Net", money(it)) }
        payload.total?.let { emitBoldKeyValue(out, "TOTAL", money(it)) }
        payload.paymentMethod?.takeIf { it.isNotBlank() }?.let { emitKeyValue(out, "Payment Method", it) }
        if (payload.payments.isNotEmpty()) {
            emitDivider(out)
            emitLineCentered(out, "PAYMENTS")
            payload.payments.forEach { payment ->
                val method = payment.method.ifBlank { payment.status ?: "Payment" }
                val amount = payment.amount?.let { money(it) } ?: ""
                emitKeyValue(out, method, amount.ifBlank { payment.reference ?: payment.status ?: "" })
                payment.reference?.takeIf { it.isNotBlank() }?.let { emitLine(out, "  Ref: $it") }
                payment.status?.takeIf { it.isNotBlank() && !it.equals("paid", ignoreCase = true) }?.let { emitLine(out, "  Status: $it") }
                payment.note?.takeIf { it.isNotBlank() }?.let { emitLine(out, "  Note: $it") }
            }
        }
        if (payload.paid != null && payload.paid > 0.0) emitKeyValue(out, "Paid", money(payload.paid))
        if (!isCreditReceipt && payload.change != null && payload.change > 0.0 && paymentStatus.equals("cash", ignoreCase = true)) emitKeyValue(out, "Change", money(payload.change))
        if (!isCreditReceipt && balanceDue > 0.0) {
            val label = if (isCreditReceipt) "BALANCE DUE" else "Balance"
            emitBoldKeyValue(out, label, money(balanceDue))
        }
        if (effectiveConditions.isNotEmpty()) {
            emitDivider(out)
            emitCenteredBoldLine(out, "CONDITIONS")
            effectiveConditions.forEachIndexed { index, condition ->
                wrapText(condition, width).forEachIndexed { partIndex, line ->
                    emitLine(out, if (partIndex == 0) "${index + 1}. $line" else "   $line")
                }
            }
        }

        emitDivider(out)
        payload.footer?.takeIf { it.isNotBlank() }?.let {
            emitCenteredBoldLine(out, it)
        } ?: emitCenteredBoldLine(out, "Thank you for shopping with us")
        emit(out, EscPos.feed(3))
        emit(out, EscPos.cut())
        return out.toByteArray()
    }

    private fun renderLogo(out: ByteArrayOutputStream) {
        try {
            val ctx = context ?: return
            val resId = ctx.resources.getIdentifier("stery_logo", "drawable", ctx.packageName)
            if (resId == 0) return
            val bitmap = BitmapFactory.decodeResource(ctx.resources, resId) ?: return
            emit(out, EscPos.alignCenter())
            emit(out, EscPos.logo(bitmap, targetWidth = 384))
            emit(out, EscPos.feed(1))
            emit(out, EscPos.alignLeft())
        } catch (_: Exception) {
            // Logo is optional.
        }
    }

    private fun formatItem(item: ReceiptItem): List<String> {
        val name = sanitize(item.name)
        Log.d("PRINT", "ANDROID ITEM NAME raw=$item resolved=$name")
        val qtyText = formatQty(item.qty)
        val unitText = money(item.unitPrice)
        val totalText = money(item.lineTotal)
        val nameWidth = ITEM_NAME_WIDTH.coerceAtMost(width - totalText.length - 1).coerceAtLeast(10)
        val nameLines = wrapText(name, nameWidth)
        val firstLeft = "${nameLines.firstOrNull().orEmpty()} $qtyText x ${unitText.removePrefix("KES ")}".trim()
        val firstLine = leftRight(firstLeft, totalText)
        val result = mutableListOf(firstLine)
        if (nameLines.size > 1) {
            nameLines.drop(1).forEach { result.add("  $it") }
        }
        return result
    }

    private fun emitLineCentered(out: ByteArrayOutputStream, text: String) {
        emit(out, EscPos.alignCenter())
        emitLine(out, text)
        emit(out, EscPos.alignLeft())
    }

    private fun emitCenteredBoldLine(out: ByteArrayOutputStream, text: String) {
        emit(out, EscPos.alignCenter())
        emit(out, EscPos.bold(true))
        emitLine(out, text)
        emit(out, EscPos.bold(false))
        emit(out, EscPos.alignLeft())
    }

    private fun emitWrappedBlock(out: ByteArrayOutputStream, text: String) {
        wrapText(text, width).forEach { emitLine(out, it) }
    }

    private fun emitKeyValue(out: ByteArrayOutputStream, left: String, right: String) {
        emitLine(out, leftRight(left, right))
    }

    private fun emitBoldKeyValue(out: ByteArrayOutputStream, left: String, right: String) {
        emit(out, EscPos.bold(true))
        emitLine(out, leftRight(left, right))
        emit(out, EscPos.bold(false))
    }

    private fun leftRight(left: String, right: String): String {
        val l = sanitize(left)
        val r = sanitize(right)
        val spaces = (width - l.length - r.length).coerceAtLeast(1)
        return if (l.length + r.length + spaces <= width) {
            l + " ".repeat(spaces) + r
        } else {
            "${l.take((width - r.length - 1).coerceAtLeast(1))} $r".trim()
        }
    }

    private fun wrapText(text: String, maxWidth: Int): List<String> {
        val words = sanitize(text).split(" ").filter { it.isNotBlank() }
        if (words.isEmpty()) return listOf("")
        val lines = mutableListOf<String>()
        var current = ""
        for (word in words) {
            val next = if (current.isBlank()) word else "$current $word"
            if (next.length > maxWidth && current.isNotBlank()) {
                lines.add(current)
                current = word
            } else {
                current = next
            }
        }
        if (current.isNotBlank()) lines.add(current)
        return lines
    }

    private fun emitDivider(out: ByteArrayOutputStream) {
        emitLine(out, "-".repeat(width))
    }

    private fun emitLine(out: ByteArrayOutputStream, text: String) {
        emit(out, EscPos.line(text))
    }

    private fun emit(out: ByteArrayOutputStream, bytes: ByteArray) {
        out.write(bytes)
    }

    private fun money(value: Double): String = "KES ${String.format(Locale.US, "%.2f", value)}"

    private fun formatQty(value: Double): String {
        return if (value == value.toInt().toDouble()) value.toInt().toString() else value.toString()
    }

    private fun sanitize(value: String): String {
        return value
            .replace("\u2018", "'")
            .replace("\u2019", "'")
            .replace("\u201C", "\"")
            .replace("\u201D", "\"")
            .replace("\u2013", "-")
            .replace("\u2014", "-")
            .replace(Regex("[^\\x09\\x0A\\x0D\\x20-\\x7E]"), "")
            .trim()
    }

    companion object {
        const val RECEIPT_WIDTH = 32
        const val ITEM_NAME_WIDTH = 16
    }
}

File: print-bridge/receiptFormatter.js
const { sanitizeText } = require("./receiptPayload");

const WIDTH = 32;

function center(text, width = WIDTH) {
  const value = sanitizeText(text);
  if (!value) return "";
  if (value.length >= width) return value.slice(0, width);
  const pad = Math.floor((width - value.length) / 2);
  return `${" ".repeat(Math.max(0, pad))}${value}`;
}

function leftRight(left, right, width = WIDTH) {
  const l = sanitizeText(left);
  const r = sanitizeText(right);
  const space = width - l.length - r.length;
  if (space >= 1) return `${l}${" ".repeat(space)}${r}`;
  const available = Math.max(1, width - r.length - 1);
  return `${l.slice(0, available)} ${r}`.trimEnd();
}

function divider(char = "-", width = WIDTH) {
  return String(char || "-").repeat(width).slice(0, width);
}

function wrapText(text, width = WIDTH) {
  const input = sanitizeText(text);
  if (!input) return [""];
  const words = input.split(/\s+/).filter(Boolean);
  const lines = [];
  let current = "";
  words.forEach((word) => {
    const next = current ? `${current} ${word}` : word;
    if (next.length > width && current) {
      lines.push(current);
      current = word;
    } else {
      current = next;
    }
  });
  if (current) lines.push(current);
  return lines.length ? lines : [input.slice(0, width)];
}

function money(value) {
  const num = Number.parseFloat(String(value ?? 0).replace(/,/g, "").replace(/[^0-9.-]/g, ""));
  if (!Number.isFinite(num)) return "KES 0.00";
  return `KES ${num.toFixed(2)}`;
}

function formatQty(value) {
  const num = Number.parseFloat(String(value ?? 0).replace(/,/g, ""));
  if (!Number.isFinite(num)) return "1";
  return Number.isInteger(num) ? String(num) : String(num);
}

function formatItem(item, width = WIDTH) {
  const name = sanitizeText(item?.name || "Item");
  const qty = formatQty(item?.qty);
  const unit = money(item?.unitPrice ?? 0);
  const total = money(item?.lineTotal ?? item?.total ?? 0);
  const left = `${name} ${qty} x ${unit.replace(/^KES\s+/i, "")}`.trim();
  const right = total.replace(/^KES\s+/i, "");
  const nameWidth = Math.max(10, width - right.length - 1);
  const wrapped = wrapText(left, nameWidth);
  const lines = [leftRight(wrapped[0] || left, right, width)];
  wrapped.slice(1).forEach((part) => lines.push(`  ${part}`));
  return lines;
}

function formatReceiptText(receipt) {
  const payload = receipt && typeof receipt === "object" ? receipt : {};
  const lines = [];
  const storeName = sanitizeText(payload.storeName || "STERY WHOLESALERS");
  const branch = sanitizeText(payload.branch || "");
  const receiptNo = sanitizeText(payload.receiptNo || "");
  const date = sanitizeText(payload.date || "");
  const cashier = sanitizeText(payload.cashierName || payload.cashier || "");
  const customer = sanitizeText(payload.customerName || payload.customer || "");
  const note = sanitizeText(payload.note || "");
  const footer = sanitizeText(payload.footer || "Thank you for shopping with us");
  const duplicateLabel = sanitizeText(payload.duplicateLabel || "");
  const paymentMethod = sanitizeText(payload.paymentMethod || "");
  const paymentStatus = sanitizeText(payload.paymentStatus || "");
  const isCredit = Boolean(payload.isCredit || /credit/i.test(paymentStatus));
  const total = payload.total ?? 0;
  const paid = payload.paid;
  const balance = payload.balance ?? payload.balanceDue ?? (isCredit ? Math.max(0, Number(total || 0) - Number(paid || 0)) : null);
  const netAmount = payload.netAmount ?? null;
  const vat = payload.vat ?? payload.tax ?? null;
  const payments = Array.isArray(payload.payments) ? payload.payments : [];
  const deliveryPerson = sanitizeText(payload.deliveryPersonName || payload.deliveryPerson || "");
  const conditions = Array.isArray(payload.conditions) ? payload.conditions.map((entry) => sanitizeText(entry)).filter(Boolean) : [];

  if (duplicateLabel) {
    lines.push(center(duplicateLabel));
  }
  lines.push(center(storeName));
  if (branch) lines.push(center(branch));
  lines.push(center("SALE RECEIPT"));
  lines.push(divider());
  if (receiptNo) lines.push(leftRight("Receipt", receiptNo));
  if (date) lines.push(leftRight("Date", date));
  if (cashier) lines.push(leftRight("Cashier", cashier));
  if (customer) lines.push(leftRight("Customer", customer));
  if (deliveryPerson && isCredit) lines.push(leftRight("Delivery Person", deliveryPerson));
  if (payload.saleType) lines.push(leftRight("Sale Type", sanitizeText(payload.saleType)));
  if (isCredit) {
    lines.push(center("CREDIT SALE"));
    if (paymentStatus && !/^(paid|cash|mpesa|card|split payment)$/i.test(paymentStatus)) {
      lines.push(leftRight("Status", paymentStatus));
    }
  } else if (paymentStatus && !/^(paid|cash|mpesa|card|split payment)$/i.test(paymentStatus)) {
    lines.push(leftRight("Status", paymentStatus));
  }
  if (note) {
    lines.push(divider());
    lines.push(...wrapText(note));
  }
  lines.push(divider());
  lines.push(leftRight("Item / Price x Qty", "Amt"));
  lines.push(divider());

  const items = Array.isArray(payload.items) ? payload.items : [];
  items.forEach((item) => {
    lines.push(...formatItem(item));
  });

  lines.push(divider());
  if (payload.subtotal !== null && payload.subtotal !== undefined) lines.push(leftRight("Subtotal", money(payload.subtotal)));
  if (Number(payload.discount || 0) > 0) {
    lines.push(leftRight("Discount", money(payload.discount)));
  }
  if (Number(vat || 0) > 0) lines.push(leftRight("VAT", money(vat)));
  if (netAmount !== null && netAmount !== undefined) lines.push(leftRight("Net", money(netAmount)));
  lines.push(leftRight("TOTAL", money(total)));
  if (paymentMethod) lines.push(leftRight("Payment Method", paymentMethod));
  if (payments.length) {
    lines.push(divider());
    lines.push(center("PAYMENTS"));
    payments.forEach((payment) => {
      const method = sanitizeText(payment.method || payment.paymentMethod || payment.type || "");
      const amount = payment.amount !== null && payment.amount !== undefined ? money(payment.amount) : "";
      const ref = sanitizeText(payment.reference || "");
      const status = sanitizeText(payment.status || "");
      const row = method || ref || status ? leftRight(method || status || "Payment", amount || ref || status) : "";
      if (row) lines.push(row);
      if (ref && method) lines.push(`  Ref: ${ref}`);
      if (status && status.toLowerCase() !== "paid") lines.push(`  Status: ${status}`);
    });
  }
  if (paid !== null && paid !== undefined && Number(paid) > 0) lines.push(leftRight("Paid", money(paid)));
  if (!isCredit && payload.change !== null && payload.change !== undefined && Number(payload.change) > 0 && /cash/i.test(paymentMethod)) {
    lines.push(leftRight("Change", money(payload.change)));
  }
  if (!isCredit && balance !== null && balance !== undefined && Number(balance) > 0) {
    lines.push(leftRight("Balance", money(balance)));
  }
  if (conditions.length) {
    lines.push(divider());
    lines.push(center("CONDITIONS"));
    conditions.forEach((condition, index) => {
      const parts = wrapText(condition);
      parts.forEach((part, partIndex) => {
        lines.push(partIndex === 0 ? `${index + 1}. ${part}` : `   ${part}`);
      });
    });
  }
  lines.push(divider());
  lines.push(center(footer));

  return lines.filter((line) => line !== undefined && line !== null).join("\n");
}

module.exports = {
  formatReceiptText,
};
